Add Standard Data Release YAML to GitHub Actions

afontani opened the pull request:
## Pull Request Description
This pull request introduces the Standard Data Release (SDR) YAML file into the "develop" branch of ResStock and into GitHub Actions integration-tests. This PR will give contributors on the SDR measures a place to commit changes and aid in completing an end-to-end workflow in CI.

### Workflow changes
1. National Upgrades and Testing Upgrades have been removed
2. Path upload errors added to `.github/workflows/config.yml`. Previously was defaulted to "warn" and wasn't catching that a file was missing. This is done by adding `if-no-files-found: error` to the `actions/upload-artifact@v4` action. Example: https://github.com/NREL/resstock/blob/037a9df8af6b800451c6d046ad32c4c50495a85f/.github/workflows/config.yml#L121-L126
3. The SDR project file tests have been added similarly to the tests for the project national and project testing upgrade files.

### Test Modifications
1. The tests have been updated to remove the tests around the National Upgrades and Testing Upgrades and switch the tests to the SDR project file.
2. In `test/test_analysis_tools.rb`: BuildStockBatch sometimes has extra blank columns. This is because of the superset of all columns from all upgrades, not just an individual upgrade. Removing these columns from the test as they are blank. https://github.com/NREL/resstock/blob/0d7fb0780b9d22c10d46f9f347c29d1042b89b40/test/test_analysis_tools.rb#L49-L54

### Area for improved testing
This PR is meant to introduce the SDR project file and do some elementary testing (similar to project national, project testing, and national upgrades). At this point, the tests make sure there is some data and columns. Improving the tests that exist today, building out functional tests, and regression testing are important areas for improvement moving forward. **Example:** using only 2 datapoints being simulated some upgrades do not get run because the upgrade is invalid for the two units.

### Related Pull Requests
None.

### Related Issues
Closes #1261 -- "Commit ResStock 2024.2 data release yaml to develop"

## Checklist

#### Required:
- [ ] [Technical reference documentation](https://github.com/NREL/resstock/tree/develop/docs/read_the_docs) has been updated
- [x] Add to the [changelog_dev.rst file](https://github.com/NREL/resstock/tree/develop/docs/read_the_docs/source/changelog/changelog_dev.rst)
- [x] No unexpected regression test changes on CI (comparison artifacts have been checked)

#### Optional (not all items may apply):
- [x] Tests (and test files) have been updated
- [ ] Supporting resource files have been updated (related to resstock-estimation)
  - [ ] [data dictionary](https://github.com/NREL/resstock/tree/develop/resources/data/dictionary)
  - [ ] [source report](https://github.com/NREL/resstock/tree/develop/project_national/resources/source_report.csv)
  - [ ] [options saturation](https://github.com/NREL/resstock/tree/develop/project_national/resources/options_saturations.csv)
  - [x] [options_lookup](https://github.com/NREL/resstock/blob/develop/resources/options_lookup.tsv)
- [x] `openstudio tasks.rb update_measures` has been run

lixiliu committed:
Update test analysis tools

Commit:
Latest results.

Commit:
Initial commit of the srd yaml file

Commit:
Add SRD yaml into integration tests

Update changelog

Add workflow generator version

update number of allowable upgrades

Add files to .gitignore

update input and output directories

Update options lookup arguments

Commit:
Add utility bills

Commit:
Update integration tests and results processing

rename sdr yaml file

Unzip and upload sdr upgrade artifact

Add testing for sdr upgrade run with bsb

Reduce upgrade datapoints on national and testing upgrades ymls

revert data point changes and reduce sdr datapoints.

Remove testing and national upgrades

Update process_bsb_analysis

Updates. test_bsb_analysis passing

Fix test_run_analysis

Update testing yml_valid

Update expected warnings

Remove testing_upgrades from gitignore

Update buildstock_missing.csv

Make sdr tests an upgrade

Add the upgrade osw and xml files as requested from SDR

Update the package name

load only headers for comparison

Add commented out production number of data points

remove Zone People Occupant Count: Conditioned Space from timeseries contents

New tsvs with updated well pump

update changelog

update tsvs

Commit:
Latest results.

joseph-robertson commented:
Remove some tests? For example, remove national/testing upgrade ymls in favor of SDR upgrade yml.

Commit:
Well pump merge changes

Increase n_datapoints

Fix option name change from well pump PR

Typo

reduce datapoints

Commit:
Latest results.

Commit:
Update analysis tests and results processing

Update test_upgrade_results

Update test_upgrade_columns

Fix paths

Fix number of expected files

Fix run_analysis file path

Fix missing upgrade results file in run_analysis results

Fix missing upgrade results file in run_analysis results, try 2

Add path errors during artifact upload

Cleanup

Update the expected number of expected results

Switch upgrade to one that applies to all units

joseph-robertson committed:
Update README.md.

Commit:
Fix path

Fix bsb sdr results csv

Commit:
Latest results.

Commit:
New tsvs with updated well pump

Commit:
Latest results.

joseph-robertson commented on project_national/sdr_upgrades_tmy3.yml:
```diff
@@ -0,0 +1,1149 @@
+  args:
+    # Annual simulation and timestep
+    build_existing_model:
+      simulation_control_timestep: 15
```
Is it necessary to have 15 instead of 60 here?

afontani replied:
We might need to for CI to get a couple more datapoints simulated. 15 min is what we run for the dataset publication.

joseph-robertson commented on project_national/sdr_upgrades_tmy3.yml:
```diff
@@ -0,0 +1,1149 @@
+
+    # Outputs to include
+    simulation_output_report:
+      timeseries_frequency: timestep
```
Can this be hourly?

afontani replied:
This is what we run for the dataset publication. I would like this to be as close as possible to the simulated version. If we forget to change this then we have to re-simulate the entire dataset. Yes, the downside is limiting the number of data points. Can we make the change later if we figure out a solution to the memory limitations?

Commit:
Merge remote-tracking branch 'origin/develop' into sdr_yaml_test

Commit:
Update changelog date

Commit:
Latest results.

Commit:
Update new results to the base_results

Commit:
remove files

joseph-robertson commented on project_national/sdr_upgrades_tmy3.yml:
```diff
@@ -0,0 +1,1149 @@
+schema_version: '0.4'
```
I'd suggest making the following updates to the baseline yml files for consistency with the new sdr_upgrades_tmy3.yml file:
* order the yml blocks the same (e.g., move `baseline` up)
* remove the `eagle` blocks

afontani replied:
Addressed: [469ce5d](https://github.com/NREL/resstock/pull/1329/commits/469ce5dea7d97864f342bb902d69471e9d35e007)

joseph-robertson commented on .github/workflows/config.yml:
```diff
@@ -192,20 +190,21 @@ jobs:
-            project_national/national_upgrades/results_csvs/results_up17.csv
-            project_testing/testing_upgrades/results_csvs/results_up17.csv
-          name: buildstockbatch_results_csvs
+            project_national/sdr_upgrades_tmy3/results_csvs/results_up16.csv
```
For the old upgrade yml files, "17" referred to the final "Package Upgrade" upgrade scenario. I believe the thinking here was that the simulation output would be the most comprehensive for testing purposes. I'm not sure that the final "16" has the same significance here for the new yml file.

Also, you may want to avoid hardcoding "16" here if the new yml will expand over time.

afontani replied:
Right. In the new SDR file, there is not an all-encompassing upgrade. Upgrade 16 is currently the "light touch envelope" measure. I used this upgrade not because it was the last upgrade, but because it applies to every unit. Expanding the number of upgrades should be fine and the file could still keep upgrade 16 here.  We are planning on testing the outputs somewhere else with the data dictionary, but that is in the future.

I am leaving this for now.

joseph-robertson commented on docs/read_the_docs/source/changelog/changelog_dev.rst:
```diff
@@ -36,3 +36,18 @@ Development Changelog
         resstock-estimation: `pull request 436 <https://github.com/NREL/resstock-estimation/pull/436>`_
 
         Assignees: Janet Reyna
```
Outside of scope of this PR, but this heat pump pool heaters entry is missing a `.. change::` line above.

afontani replied:
Resolved in https://github.com/NREL/resstock/pull/1338

joseph-robertson commented on measures/ApplyUpgrade/resources/constants.rb:
```diff
@@ -1,7 +1,7 @@
 
 module Constants
-  NumApplyUpgradeOptions = 25
+  NumApplyUpgradeOptions = 100
```
This requires an update to RTD at docs/read_the_docs/advanced_tutorial/increasing_upgrade_options.rst.

afontani replied:
Addressed: [91cdc87](https://github.com/NREL/resstock/pull/1329/commits/91cdc876c3c5b6ee45337aae5eb294062f4c138a)

joseph-robertson commented on docs/read_the_docs/source/workflow_inputs/characteristics.rst:
```diff
@@ -40097,6 +40097,12 @@ Arguments
      - Double
      -
      - The maximum capacity limit applied to the auto-sizing methodology. If not provided, no limit is used.
+   * - ``heating_system_2_fraction_heat_load_served``
```
This argument is being added here under "HVAC Secondary Heating Efficiency" because you've added new "Fuel Wall/Floor Furnace" options which assign `heating_system_2_fraction_heat_load_served`. But that argument is already assigned in "HVAC Secondary Heating Partial Space Conditioning"...

afontani replied:
I think this will fix this copy error. [3ccb1cd](https://github.com/NREL/resstock/pull/1329/commits/3ccb1cd000da769878d212e17b8bbea6ae587a84)

joseph-robertson commented on resources/data/utility_bills/sdr_rates/README.md:
```diff
@@ -0,0 +1,6 @@
+In the YML file's "simple_filepath" field for utility bill scenario definitions, enter a relative file path to a TSV lookup file (e.g., "data/simple_rates/State.tsv") containing user-defined values corresponding to arguments for fixed costs, marginal rates, and PV.
+The first column of the TSV lookup file contains the name of a chosen parameter for which sets of argument values are assigned according to its options.
+Any blank fields, or missing options for a parameter not specified in the TSV lookup file, will be defaulted.
+
+See the Simple section of OpenStudio-HPXML's documentation for [Electricity Rates](https://openstudio-hpxml.readthedocs.io/en/latest/workflow_inputs.html#electricity-rates), and [Fuel Rates](https://openstudio-hpxml.readthedocs.io/en/latest/workflow_inputs.html#fuel-rates), for more information about arguments for fixed charges, marginal rates, and PV.
+Refer to BuildStockBatch's documentation for [Residential HPXML Workflow Generator](https://buildstockbatch.readthedocs.io/en/stable/workflow_generators/residential_hpxml.html) for more information about YML file arguments.
```
This appears to be copied from resources/data/utility_bills/README.md. Instead, cite the source of the data contained in sdr_rates/State.tsv?

afontani replied:
Good catch. Addressed: [2cae49a](https://github.com/NREL/resstock/pull/1329/commits/2cae49a089acd457f11618d20d7cfac1310fcbc9)

joseph-robertson commented on test/test_bsb_analysis.rb:
```diff
@@ -10,6 +10,7 @@ def before_setup
     @national_baseline = 'project_national/national_baseline'
     @testing_upgrades = 'project_testing/testing_upgrades'
     @national_upgrades = 'project_national/national_upgrades'
```
Delete these.

afontani replied:
Oops. Addressed: [a46810d](https://github.com/NREL/resstock/pull/1329/commits/a46810da424efc9f381f3a52aa4c71a14282efc4)

joseph-robertson commented on test/test_run_analysis.rb:
```diff
@@ -15,14 +15,13 @@ def before_setup
 
     @testing_baseline = File.join(buildstock_directory, 'testing_baseline')
-    @national_baseline = File.join(buildstock_directory, 'national_baseline')
     @testing_upgrades = File.join(buildstock_directory, 'testing_upgrades')
```
I was going to suggest deleting this too, but appears its used in some tests below. Suggest a name change so that it's not confused with the "testing project"? Maybe that's overkill; your call.

afontani replied:
I moved the `@testing_upgrades` variable down to the test where the variable is called since it is the only time the variable is used. I also renamed the output folder in the yaml file. [082c1b5](https://github.com/NREL/resstock/pull/1329/commits/082c1b55020524f4ef1a7f67631453ffc0b2a879) and [4384f89](https://github.com/NREL/resstock/pull/1329/commits/4384f89338e0316af1275894f61f26692457c500)

Commit:
Merge remote-tracking branch 'origin' into sdr_yml

Commit:
Make yml structure consistent

Commit:
Update TDG

Commit:
Correct heating_system_2_fraction_heat_load_served specification

Commit:
Update utility rates README

Commit:
Remove unused variables

Commit:
Update testing_upgrades variable

Commit:
testing_upgrade, try 2

Commit:
One more fix

Commit:
Latest results.

Commit:
clean up

Commit:
Update changelog

joseph-robertson commented on resources/data/utility_bills/sdr_rates/State.tsv:
```diff
@@ -0,0 +1,50 @@
+State	electricity_fixed_charge	electricity_marginal_rate	natural_gas_fixed_charge	natural_gas_marginal_rate	propane_fixed_charge	propane_marginal_rate	fuel_oil_fixed_charge	fuel_oil_marginal_rate	wood_fixed_charge	wood_marginal_rate
```
Ugh. Missed this in the original add of this file. These electricity columns should be "elec_fixed_charge" and "elec_marginal_rate", otherwise these fields will be ignored and instead populated with OS-HPXML defaults.

@afontani @jbrossman
